clang/lib/CIR/CodeGen/CIRGenStmtOpenMP.cpp

```diff
@@ -43,3 +43,35 @@ CIRGenFunction::buildOMPParallelDirective(const OMPParallelDirective &S) {
   builder.create<TerminatorOp>(getLoc(S.getSourceRange().getEnd()));
   return res;
 }
+
+mlir::LogicalResult
+CIRGenFunction::buildOMPTaskwaitDirective(const OMPTaskwaitDirective &S) {
+  mlir::LogicalResult res = mlir::success();
```

Comment:
Looking at the traditional LLVM codegen in Clang, I see:
```
  OMPTaskDataTy Data;
  // Build list of dependences
  buildDependences(S, Data);
  Data.HasNowaitClause = S.hasClausesOfKind<OMPNowaitClause>();
  CGM.getOpenMPRuntime().emitTaskwaitCall(*this, S.getBeginLoc(), Data);
```

For the most part, the CIRGen code should look very similar to this, with the core of `buildDependences` and `emitTaskwaitCall` doing the actual MLIR based openMP support in CIR. A good start is to add a simple version of `buildDependences` and implement the minimal `CIRGenOpenMPRuntime` needed for a `emitTaskwaitCall` to build the omp dialect operations.

Reply:
Okey now I'm understanding the underlying issue of my implementation. OpenMP has a runtime library that has to perform some checks before emitting certain instructions like taskwait. To continue the implementation I wanted to ask you If it would be a better idea to reimplement data structures like OMPTaskDataTy on ClangIR or to just include them. I would personally include those said structures, but I'm not sure at all. 

Also I wanted to point out that almost every PR's is bugged for some reason.
![image](https://github.com/llvm/clangir/assets/81811777/da344d0b-d977-4440-a6e0-6c5e41b9df7a)

Reply:
> To continue the implementation I wanted to ask you If it would be a better idea to reimplement data structures like OMPTaskDataTy on ClangIR or to just include them

Reimplement, CIR doesn't have access to `lib/CodeGen`. The idea is to add the minimal skeleton for `OMPTaskDataTy` you need for this PR, and for other data structures and pieces you might need for `CGOpenMPRuntime.h`. The idea is not to blindly copy it, but add the pieces incrementally in each PR, as the need appears.

> Do i add the CGOpenMPRuntime.h to the clang/include dir, or do I rewrite the data structures from scratch inside ClangIR?

Rewrite them, we already do that a bunch, so you can look at the existing stuff for inspiration.

> Also I wanted to point out that almost every PR's is bugged for some reason

This happens when we rebase the clangir repo against LLVM upstream. You need to rebase your local work against our repo and force push the PR to fix it.

Reply:
I finally got to rebase the branch correctly, but now It seems that there is some duplicity on the commits. I'm trying to address this issue using git rebase -i and squashing the commits together but for some reason I can't do that without getting merge conflicts. Appart from that, could you @bcardosolopes review the last minor changes I've uploaded for OpenMPRuntime? I kept them to minimals except for buildDependences, which I thought that was convenient to use the original Clang one.

Reply:
Interesting, not sure what's going on - people do it all the time and it usually works, make sure you are rebasing against clangir's repo upstream, not LLVM's main repo upstream.

If you cannot make the rebase happen, feel free to start a new PR (after you apply the requested changes).

> Appart from that, could you @bcardosolopes review the last minor changes I've uploaded for OpenMPRuntime?

I rather look into a clean PR instead.